esphome/components/modbus_controller/select/modbus_select.cpp

```diff
@@ -0,0 +1,89 @@
+#include "modbus_select.h"
+#include "esphome/core/log.h"
+
+namespace esphome {
+namespace modbus_controller {
+
+static const char *const TAG = "modbus_controller.select";
+
+void ModbusSelect::dump_config() { LOG_SELECT(TAG, "Modbus Controller Select", this); }
+
+void ModbusSelect::parse_and_publish(const std::vector<uint8_t> &data) {
+  int64_t value = payload_to_number(data, this->sensor_value_type, this->offset, this->bitmask);
+
+  ESP_LOGD(TAG, "New select value %lld from payload", value);
+
+  optional<std::string> new_state;
+
+  if (this->transform_func_.has_value()) {
+    auto val = (*this->transform_func_)(this, value, data);
+    if (val.has_value()) {
+      new_state = *val;
+      ESP_LOGV(TAG, "lambda returned option %s", new_state->c_str());
+    }
+  }
+
+  if (!new_state.has_value()) {
+    auto map_it = std::find(this->mapping_.cbegin(), this->mapping_.cend(), value);
+
+    if (map_it != this->mapping_.cend()) {
+      size_t idx = std::distance(this->mapping_.cbegin(), map_it);
+      new_state = this->traits.get_options()[idx];
+      ESP_LOGV(TAG, "Found option %s for value %lld", new_state->c_str(), value);
+    } else {
+      ESP_LOGE(TAG, "No option found for mapping %lld", value);
+    }
+  }
+
+  if (new_state.has_value()) {
+    this->publish_state(new_state.value());
+  }
+}
+
+void ModbusSelect::control(const std::string &value) {
+  optional<int64_t> mapval;
+  std::vector<uint16_t> data;
+
+  if (this->write_transform_func_.has_value()) {
+    auto val = (*this->write_transform_func_)(this, value, data);
+    if (val.has_value()) {
+      mapval = *val;
+      ESP_LOGV(TAG, "write_lambda returned mapping value %lld", *mapval);
+    } else {
+      ESP_LOGD(TAG, "Communication handled by write_lambda - exiting control");
+      return;
+    }
+  }
+
+  if (data.empty()) {
+    if (!mapval.has_value()) {
+      auto options = this->traits.get_options();
+      auto opt_it = std::find(options.cbegin(), options.cend(), value);
+      size_t idx = std::distance(options.cbegin(), opt_it);
+      mapval = this->mapping_[idx];
+      ESP_LOGV(TAG, "Found value %lld for option '%s'", *mapval, value.c_str());
+    }
+
+    number_to_payload(data, *mapval, this->sensor_value_type);
```

Comment:
```suggestion
#pragma GCC diagnostic push
#pragma GCC diagnostic ignored "-Wmaybe-uninitialized"
    number_to_payload(data, *mapval, this->sensor_value_type);
#pragma GCC diagnostic pop
```
I'm not sure why a `maybe-uninitialized`  warning is thrown here. Unless someone else sees a logic error I suggest disabling it for this line. 

Users without coding experience often misread a warning as an indicator that something is broken in esphome

Reply:
Agreed. I had this pragma already in place, but a check fails then. Let's see what happens now.

Reply:
Check failed again:

```
Error: /home/runner/work/esphome/esphome/esphome/components/modbus_controller/select/modbus_select.cpp:68:32: error: unknown warning group '-Wmaybe-uninitialized', ignored [clang-diagnostic-unknown-warning-option,-warnings-as-errors]
#pragma GCC diagnostic ignored "-Wmaybe-uninitialized"
```

Any ideas?

Reply:
A workaround for a workaround :)
At least  locally running lint from docker works with this 
```
#pragma GCC diagnostic push
#if !defined( __clang__ )
#pragma GCC diagnostic ignored "-Wmaybe-uninitialized"
#endif
```

Reply:
I thought a lot about your suggestion of `write_lambda`. I've got another idea which should be compatible with the existing modbus write lambdas. Please check it.

This change seems also to fix the false-positive warning about uninitialized variables.

Reply:
Looks good to me